node/node.go

```diff
@@ -0,0 +1,452 @@
+		return nil, nil, err
+	}
+	walletAddr := wallet.StandardAddress(walletKey.PublicKey())
+	ws, ccid, err := stores.NewJSONWalletStore(walletDir, walletAddr)
+	if err != nil {
+		return nil, nil, err
+	} else if err := cm.ConsensusSetSubscribe(ws, ccid, nil); err != nil {
+		return nil, nil, err
+	}
+	w := wallet.NewSingleAddressWallet(walletKey, ws)
+
+	dbDir := filepath.Join(dir, "db")
+	if err := os.MkdirAll(dbDir, 0700); err != nil {
+		return nil, nil, err
+	}
+	dbConn := stores.NewSQLiteConnection(filepath.Join(dbDir, "db.sqlite"))
+
+	hostdbDir := filepath.Join(dir, "hostdb")
+	if err := os.MkdirAll(hostdbDir, 0700); err != nil {
+		return nil, nil, err
+	}
+	sqlStore, ccid, err := stores.NewSQLStore(dbConn, true)
+	if err != nil {
+		return nil, nil, err
+	} else if err := cm.ConsensusSetSubscribe(sqlStore, ccid, nil); err != nil {
+		return nil, nil, err
+	}
+
+	contractsDir := filepath.Join(dir, "contracts")
+	if err := os.MkdirAll(contractsDir, 0700); err != nil {
+		return nil, nil, err
+	}
+
+	objectsDir := filepath.Join(dir, "objects")
+	if err := os.MkdirAll(objectsDir, 0700); err != nil {
+		return nil, nil, err
+	}
+
+	cleanup := func() error {
+		errs := []error{
+			g.Close(),
+			cm.Close(),
+			tp.Close(),
+			sqlStore.Close(),
+		}
+		for _, err := range errs {
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	b := bus.New(syncer{g, tp}, chainManager{cm}, txpool{tp}, w, sqlStore, sqlStore, sqlStore, sqlStore)
+	return b, cleanup, nil
+}
+
+func newWorker(cfg WorkerConfig, walletKey consensus.PrivateKey) (*worker.Worker, func() error, error) {
+	b := bus.NewClient(cfg.BusAddr, cfg.BusPassword)
+	workerKey := blake2b.Sum256(append([]byte("worker"), walletKey...))
+	w := worker.New(workerKey, b)
+	return w, func() error { return nil }, nil
+}
+
+func newAutopilot(cfg AutopilotConfig, dir string) (*autopilot.Autopilot, func() error, error) {
+	store, err := stores.NewJSONAutopilotStore(dir)
+	if err != nil {
+		return nil, nil, err
+	}
+	b := bus.NewClient(cfg.BusAddr, cfg.BusPassword)
+	w := worker.NewClient(cfg.WorkerAddr, cfg.WorkerPassword)
+	a, err := autopilot.New(store, b, w, cfg.LoopInterval)
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanup := func() error {
+		a.Stop()
+		return nil
+	}
+	return a, cleanup, nil
+}
+
+// Node describes a single renterd process with one or more enabled components.
+// A node always needs to either create a bus or connect to an existing one by
+// calling CreateBus or AddBus respectively.
+// Afterwards a worker can be created or added with CreateWorker/AddWorker. If
+// the worker is created, it will automatically connect to the known bus. If it
+// is added, the caller needs to guarantee that the worker is connected to the
+// same bus as specified in the node.
+// After both a bus and a worker were added, the autopilot can be created. It
+// will connect to the bus and worker from the previous steps.
+type Node struct {
```

Comment:
I figured a `Node` would be essentially be an initialised server listening on a specific endpoint. So we would create a `bus`, `autopilot` and `worker` nodes which would be simple wrappers around their respective clients. The autopilot would need a bus in the constructor but workers would be added dynamically. In a single node setup the servers would in fact be a single server that exposes all client APIs.

I think I just don't understand why we need `AddBus` and `CreateBus` type methods that error out when we find they've already been set. Perhaps we should have `Node` wrap clients along with a cleanup function and the endpoint url, and introduce a `Cluster` that is very similar to what you have now in `TestCluster`

Reply:
I won't write it out in full but I gave it 5 minutes so you could get an idea of what I had in mind.

```go
type AutoPilot interface {
	AddWorker(n *Node) error
	SetBus(n *Node) error
}

type Worker interface {
	SetAutopilot(n *Node) error
	SetBus(n *Node) error
}

type Node struct {
	kind     string

	client interface{}
	server http.Handler

	startFn func() error
	stopFn  func(ctx context.Context) error
}

func newAutopilot(endpoint, password string) *Node {
	return &Node{
		kind:     "autopilot",
		client:   autopilot.NewClient(endpoint, password),
		server:   jape.BasicAuth(password)(autopilot.NewServer(&autopilot.Autopilot{})),
	}
}

func newBus(endpoint, password string) *Node {
	return &Node{
		kind:     "bus",
		client:   bus.NewClient(endpoint, password),
		server:   jape.BasicAuth(password)(bus.NewServer(&bus.Bus{})),
	}
}

func newWorker(endpoint, password string) *Node {
	return &Node{
		kind:     "worker",
		client:   worker.NewClient(endpoint, password),
		server:   jape.BasicAuth(password)(worker.NewServer(&worker.Worker{})),
	}
}

type Cluster struct {
	ap      *Node
	bus     *Node
	workers []*Node
}

func newTestCluster() *Cluster {
	c := &Cluster{}
	c.addNode(newAutopilot("127.0.0.1:0", "appassword"))
	c.addNode(newBus("127.0.0.1:1", "buspassword"))
	c.addNode(newWorker("127.0.0.1:2", "workerpassword1"))
	c.addNode(newWorker("127.0.0.1:3", "workerpassword2"))
	c.addNode(newWorker("127.0.0.1:4", "workerpassword3"))
	return c
}

func (c *Cluster) start() error {
	// call all start fn
}

func (c *Cluster) stop(ctx context.Context) error {
	// call all stop fn
}

func (c *Cluster) addNode(n *Node) error {
	switch n.kind {
	case "autopilot":
		if c.ap != nil {
			return errors.New("cluster can only have one autopilot")
		}
		c.ap = n
		if c.bus != nil {
			return c.ap.client.(AutoPilot).SetBus(c.bus)
		}
		for _, w := range c.workers {
			if err := w.client.(Worker).SetAutopilot(n); err != nil {
				return err
			}
		}
	case "bus":
		if c.bus != nil {
			return errors.New("cluster can only have one autopilot")
		}
		c.bus = n
		if c.ap != nil {
			return c.ap.client.(AutoPilot).SetBus(c.bus)
		}
		for _, w := range c.workers {
			if err := w.client.(Worker).SetBus(n); err != nil {
				return err
			}
		}
	case "worker":
		c.workers = append(c.workers, n)
		if c.ap != nil {
			if err := n.client.(Worker).SetAutopilot(c.ap); err != nil {
				return err
			}
		}
		if c.bus != nil {
			if err := n.client.(Worker).SetBus(c.bus); err != nil {
				return err
			}
		}
	}
	return nil
}
```